shared/controlflow/codeql/controlflow/Cfg.qll

```diff
@@ -1514,6 +1528,260 @@ module MakeWithSplitting<
     /** Holds if CFG scope `scope` lacks an initial AST node. */
     query predicate scopeNoFirst(CfgScope scope) { not scopeFirst(scope, _) }
   }
+
+  /**
+   * Provides a basic block construction on top of the control flow graph.
+   */
+  module BasicBlocks {
+    private import codeql.controlflow.BasicBlock as BB
+
+    private class NodeAlias = Node;
+
+    private module BasicBlockInputSig implements BB::InputSig<Location> {
+      class SuccessorType = Input::SuccessorType;
+
+      predicate successorTypeIsCondition = Input::successorTypeIsCondition/1;
+
+      class Node = NodeAlias;
+
+      Node nodeGetASuccessor(Node node, SuccessorType t) { result = node.getASuccessor(t) }
+
+      predicate nodeIsDominanceEntry(Node node) { node instanceof EntryNode }
+
+      predicate nodeIsPostDominanceExit(Node node) { node.(AnnotatedExitNode).isNormal() }
+    }
+
+    private module BasicBlockImpl = BB::Make<Location, BasicBlockInputSig>;
+
+    /**
+     * A basic block, that is, a maximal straight-line sequence of control flow nodes
+     * without branches or joins.
+     */
+    final class BasicBlock extends BasicBlockImpl::BasicBlock {
+      // We extend `BasicBlockImpl::BasicBlock` to add the `getScope`.
+      /** Gets the scope of this basic block. */
+      CfgScope getScope() {
+        if this instanceof EntryBasicBlock
+        then result = this.getFirstNode().getScope()
+        else result = this.getAPredecessor().getScope()
```

Comment:
Why is this recursive and not just the following?
```suggestion
        result = this.getFirstNode().getScope()
```

Reply:
I have been wondering the same, it's hard to imagine how they would not be equivalent for correct implementations of `getScope` and `EntryBasicBlock`, but I simply reproduced the existing code (which used overloading but was otherwise the same) thinking that probably was some reason 😅

Reply:
You're right: `result = this.getFirstNode().getScope()` should be equivalent for all correct implementations.